File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilNagumoCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil;

import br.com.lett.crawlernode.core.fetcher.ProxyCollection;
import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.crawlers.extractionutils.ranking.BrasilSitemercadoCrawler;
import br.com.lett.crawlernode.util.CrawlerUtils;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BrasilNagumoCrawler extends BrasilSitemercadoCrawler {

   private Map<String, Integer> lojaInfo = getLojaInfo();
   private Double latitude = session.getOptions().optDouble("latitude");
   private Double longitude = session.getOptions().optDouble("longitude");

   public BrasilNagumoCrawler(Session session) {
      super(session);
   }

   @Override
   protected String apiSearchUrl(String lojaId) {
      return "https://www.nagumo.com.br/api/b2c/product?store_id=" + lojaId + "&text=";
   }

   @Override
   protected Map<String, Integer> getLojaInfo() {
      return super.getLojaInfo();
   }

   @Override
   protected JSONObject crawlProductInfo() {
      String apiUrl = apiSearchUrl(lojaInfo.get("IdLoja").toString()) + this.keywordEncoded.replace("+", "%20");

      Map<String, String> headers = new HashMap<>();
      headers.put("hosturl", "www.nagumo.com.br");
      headers.put("Accept", "application/json, text/plain, */*");
      headers.put("sm-token", "{\"Location\":{\"Latitude\":" + latitude + ",\"Longitude\":" + longitude + "},\"IdLoja\":" + lojaInfo.get("IdLoja") + ",\"IdRede\":" + lojaInfo.get("IdRede") + "}");

      Request requestApi = Request.RequestBuilder.create()
         .setUrl(apiUrl)
         .setProxyservice(List.of(ProxyCollection.BUY, ProxyCollection.NETNUT_RESIDENTIAL_BR, ProxyCollection.LUMINATI_RESIDENTIAL_BR))
         .setCookies(cookies)
         .setHeaders(headers)
         .build();

      return CrawlerUtils.stringToJson(this.dataFetcher.get(session, requestApi).getBody());
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/brasil/BrasilNagumoCrawler.kt
package br.com.lett.crawlernode.crawlers.ranking.keywords.brasil

import br.com.lett.crawlernode.core.fetcher.models.Request
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.crawlers.extractionutils.ranking.BrasilSitemercadoCrawler
import br.com.lett.crawlernode.util.CommonMethods
import br.com.lett.crawlernode.util.CrawlerUtils
import br.com.lett.crawlernode.util.JSONUtils
import org.json.JSONObject

class BrasilNagumoCrawler(session: Session) : BrasilSitemercadoCrawler(session) {
   companion object {
      private const val API_URL ="https://b2c-api-premiumlabel-production.azurewebsites.net/api/v1/b2c/"
   }

   override fun getHomePage(): String {
      val HOME_PAGE = session.options.optString("url")

      return HOME_PAGE
   }

   override fun crawlProductInfo(): JSONObject? {
      val lojaUrl = CommonMethods.getLast(homePage.split("www.nagumo.com.br").toTypedArray())
      val loadUrl = API_URL + "page/store" + lojaUrl
      var lojaId = ""
      var lojaRede: String? = ""
      val headers: MutableMap<String, String> = HashMap()
      headers["referer"] = homePage
      headers["accept"] = "application/json, text/plain, */*"
      headers["content-type"] = "application/json"
      val request = Request.RequestBuilder.create()
         .setUrl(loadUrl).setCookies(cookies)
         .setHeaders(headers)
         .setPayload(loadPayload)
         .build()
      val response = dataFetcher[session, request]
      val responseHeaders = response.headers
      if (responseHeaders.containsKey("sm-token")) {
         val header = responseHeaders["sm-token"]
         val token = JSONObject(header)
         lojaId = Integer.toString(JSONUtils.getIntegerValueFromJSON(token, "IdLoja", 0))
         if (lojaId == "0") {
            val body = JSONObject(response.body)
            lojaId = (body.optQuery("/sale/id") as Int).toString()
            lojaRede =(body.optQuery("/sale/idRede") as Int).toString()
            token.put("IdLoja", lojaId)
            token.put("IdRede", lojaRede)
         }
         headers["sm-token"] = token.toString()
      }
      val apiUrl = "${apiSearchUrl(lojaId)}&text=$keywordEncoded"

      val requestApi = Request.RequestBuilder.create()
         .setUrl(apiUrl)
         .setCookies(cookies)
         .setHeaders(headers)
         .build()
      return CrawlerUtils.stringToJson(dataFetcher[session, requestApi].body)
   }

   override fun getApiUrl(): String {
      return API_URL
   }
   override fun apiSearchUrl(lojaId: String): String? {
      return "https://b2c-api-premiumlabel-production.azurewebsites.net/api/b2c/product?store_id=$lojaId"
   }
}
